operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/event/source/informer/InformerEventSource.java

```diff
@@ -11,87 +13,218 @@
 import io.javaoperatorsdk.operator.api.reconciler.EventSourceContext;
 import io.javaoperatorsdk.operator.processing.event.Event;
 import io.javaoperatorsdk.operator.processing.event.EventHandler;
+import io.javaoperatorsdk.operator.processing.event.ResourceID;
 import io.javaoperatorsdk.operator.processing.event.source.ResourceCache;
 
-public class InformerEventSource<T extends HasMetadata, P extends HasMetadata>
-    extends ManagedInformerEventSource<T, P, InformerConfiguration<T, P>>
-    implements ResourceCache<T>, ResourceEventHandler<T> {
+public class InformerEventSource<R extends HasMetadata, P extends HasMetadata>
+    extends ManagedInformerEventSource<R, P, InformerConfiguration<R, P>>
+    implements ResourceCache<R>, ResourceEventHandler<R> {
+
+  private static final Logger log = LoggerFactory.getLogger(InformerEventSource.class);
 
-  private final InformerConfiguration<T, P> configuration;
+  private final InformerConfiguration<R, P> configuration;
+  private final EventBuffer<R> eventBuffer = new EventBuffer<>();
+  private final ReentrantLock lock = new ReentrantLock();
 
-  public InformerEventSource(InformerConfiguration<T, P> configuration,
-      EventSourceContext<P> context) {
+  public InformerEventSource(
+      InformerConfiguration<R, P> configuration, EventSourceContext<P> context) {
     super(context.getClient().resources(configuration.getResourceClass()), configuration);
     this.configuration = configuration;
   }
 
-  public InformerEventSource(InformerConfiguration<T, P> configuration,
-      KubernetesClient client) {
+  public InformerEventSource(InformerConfiguration<R, P> configuration, KubernetesClient client) {
     super(client.resources(configuration.getResourceClass()), configuration);
     this.configuration = configuration;
   }
 
   @Override
-  public void onAdd(T t) {
-    propagateEvent(t);
+  public void onAdd(R resource) {
+    lock.lock();
+    try {
+      var resourceID = ResourceID.fromResource(resource);
+      if (eventBuffer.isEventsRecordedFor(resourceID)) {
+        eventBuffer.eventReceived(resource);
+        return;
+      }
+      if (temporalCacheHasResourceWithVersionAs(resource)) {
+        super.onAdd(resource);
+        log.debug(
+            "Skipping event propagation for Add, resource with same version found in temporal cache: {}",
```

Comment:
Better to explain what this means for events rather than talk about the cache here

Reply:
changed it, although this is a debug log, so mostly for debugging for developers, where both are benefitial.